File: yet/0703.py
# TLE

class Node:
    def __init__(self, val):
        self.val = val
        self.next = None

class KthLargest:
    def __init__(self, k: int, nums: List[int]):
        self.k = k

        nums.sort(reverse=True)
        if len(nums) == 0:
            self.head = None
        else:
            self.head = Node(nums[0])
        tmp = self.head
        for item in nums[1:]:
            next = Node(item)
            tmp.next = next
            tmp = next
        
    def add(self, val: int) -> int:
        if self.head == None:
            self.head = Node(val)
        elif self.head.val <= val:
            n = Node(val)
            n.next = self.head
            self.head = n
        elif self.head.next == None:
                n = Node(val)
                self.head.next = n
        else:
            tmp = self.head
            while True:
                if tmp.next == None:
                    n = Node(val)
                    tmp.next = n
                    break
                elif tmp.val >= val and val >= tmp.next.val:
                    n = Node(val)
                    n.next = tmp.next
                    tmp.next = n
                    break
                tmp = tmp.next

        # linked listの表示
        # tmp = self.head
        # while tmp != None:
        #     print(tmp.val, "->", end="")
        #     tmp = tmp.next
        # print()
        
        tmp = self.head
        for i in range(self.k - 1):
            tmp = tmp.next
        return tmp.val

                

File: solved/0703.go
type Node struct {
	val int
	next *Node
}

type KthLargest struct {
	K int
	head *Node
	elemNum int
}

func makeLinkedList(nums []int) *Node {
	if len(nums) == 0 {
		return nil
	}

	// numsをソートする
	sort.Sort(sort.Reverse(sort.IntSlice(nums)))

	head := &Node{val: nums[0]}
	cur := head

	for i := 1; i < len(nums); i++ {
		tmp := &Node{val: nums[i], next: nil}
		cur.next = tmp
		cur = tmp
	}
	return head
}

func (this *KthLargest) appendList(newVal int) {
	newNode := &Node{val: newVal, next: nil}
	if this.head == nil {
		this.head = newNode
		this.elemNum++
		return
	}

	// 先頭に挿入
	if this.head.val <= newVal {
		newNode.next = this.head
		this.head = newNode
		this.elemNum++
		return
	}

	// 中間に挿入
	pivot := this.head
	for i := 0; i < this.elemNum - 1; i++ {
		if pivot.next.val <= newVal && newVal <= pivot.val {
			newNode.next = pivot.next
			pivot.next = newNode
			this.elemNum++
			return
		}
		pivot = pivot.next
	}

	// 末尾に挿入
	pivot.next = newNode	

	this.elemNum++
}

func Constructor(k int, nums []int) KthLargest {
	head := makeLinkedList(nums)
	KL := KthLargest{K: k, head: head, elemNum: len(nums)}
	return KL
}

func (this *KthLargest) printList() {
	for node := this.head; node != nil; node = node.next {
		fmt.Printf(" %d ->", node.val)
	}
	fmt.Println()
}

func (this *KthLargest) Add(val int) int {
	this.appendList(val)
	
	node := this.head
	for i := 1; i < this.K; i++ {
		node = node.next
	}
	return node.val
}


/**
 * Your KthLargest object will be instantiated and called as such:
 * obj := Constructor(k, nums);
 * param_1 := obj.Add(val);
 */